File: Modules/Module3/ChitChat/app/src/main/java/com/example/chitchat/GroupChatActivity.java
package com.example.chitchat;

import androidx.annotation.AnyThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import Adapters.MessageAdapter;
import Models.AllMethods;
import Models.ChatRoom;
import Models.Message;
import Models.Message.MessageType;
import Models.MessageRes;
import Models.User;

public class GroupChatActivity extends AppCompatActivity implements View.OnClickListener {


    FirebaseAuth auth;
    FirebaseDatabase database;
    DatabaseReference reference;
    MessageAdapter adapter;
    User u;
    ArrayList<Message> messages;

    RecyclerView rvMessage;
    ObjectMapper objectMapper = new ObjectMapper();
    EditText uInput;
    ImageButton sendBtn;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_group_chat);

        init();

    }

    private void  init( ){
        auth = FirebaseAuth.getInstance();
        database = FirebaseDatabase.getInstance();
        u = new User();

        rvMessage = findViewById(R.id.messages);
        uInput = findViewById(R.id.userMessage);
        sendBtn = findViewById(R.id.SendButton);
        sendBtn.setOnClickListener(this);
        messages = new ArrayList<>();
    }

    @Override
    public void onClick(View v) {
        if(!TextUtils.isEmpty(uInput.getText())){
            Date currentDate = new Date();

            // Format the date and time
            SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());
            String formattedDateTime = dateFormat.format(currentDate);
            Message message = new Message(uInput.getText().toString(), formattedDateTime,u.getName(),AllMethods.uId, Message.MessageType.Text);
            uInput.setText("");

            reference.push().setValue(message);
        }
    }

    @Override
    protected void onStart() {
        super.onStart();

        final FirebaseUser firebaseUser = auth.getCurrentUser();
        u.setUid(firebaseUser.getUid());
        u.setEmail(firebaseUser.getEmail());

        database.getReference("Users").child(firebaseUser.getUid()).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                u = snapshot.getValue(User.class);
                u.setUid(firebaseUser.getUid());
                AllMethods.name = u.getName();
                AllMethods.uId = firebaseUser.getUid();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });


        reference = database.getReference("chat_rooms").child(AllMethods.CurrentSelectedRoom.getKey()).child("messages");
        reference.addChildEventListener(new ChildEventListener() {
            @Override
            public void onChildAdded(@NonNull DataSnapshot snapshot, @Nullable String previousChildName) {
//                ArrayList<HashMap<String,Object>> m = (ArrayList<HashMap<String, Object>>) snapshot.getValue();
//                messages = new ArrayList<>();
//                for (HashMap<String,Object> map:m
//                     ) {
//                    messages.add(objectMapper.convertValue(map,Message.class));
//                }
//                displayMessages(messages);
                Message m = snapshot.getValue(Message.class);
                messages.add(m);
                displayMessages(messages);
            }

            @Override
            public void onChildChanged(@NonNull DataSnapshot snapshot, @Nullable String previousChildName) {
//                ArrayList<HashMap<String,Object>> m = (ArrayList<HashMap<String, Object>>) snapshot.getValue();
//                int i = 0,count = 0;
//                for (Message map:messages
//                ) {
//                    Message newM =objectMapper.convertValue(map,Message.class);
//                    if(map.getKey().equals(newM.getKey())){
//                        count = i;
//                        break;
//                    }
//                    i++;
//                }
//                messages.remove(count);
//                assert m != null;
//                messages.add(count,objectMapper.convertValue(m.get(0),Message.class));
//
//                displayMessages(messages);


//                ArrayList<HashMap<String,Object>> m = (ArrayList<HashMap<String, Object>>) snapshot.getValue();
//                messages = new ArrayList<>();
//                for (HashMap<String,Object> map:m
//                ) {
//                    messages.add(objectMapper.convertValue(map,Message.class));
//                }
//                displayMessages(messages);

                int i = 0,count = 0;
                Message m = snapshot.getValue(Message.class);
                for (Message marr : messages){
                    if(marr.getKey().equals(m.getKey())){
                        count = i;
                    }
                    i++;
                }
                messages.remove(count);
                messages.add(count,m);
            }

            @Override
            public void onChildRemoved(@NonNull DataSnapshot snapshot) {

//                ArrayList<HashMap<String,Object>> m = (ArrayList<HashMap<String, Object>>) snapshot.getValue();
//                int i = 0,count = 0;
//                for (Message map:messages
//                ) {
//                    Message newM =objectMapper.convertValue(map,Message.class);
//                    if(map.getKey().equals(newM.getKey())){
//                        count = i;
//                        break;
//                    }
//                    i++;
//                }
//                messages.remove(count);
//                displayMessages(messages);


//                ArrayList<HashMap<String,Object>> m = (ArrayList<HashMap<String, Object>>) snapshot.getValue();
//                messages = new ArrayList<>();
//                for (HashMap<String,Object> map:m
//                ) {
//                    messages.add(objectMapper.convertValue(map,Message.class));
//                }
//                displayMessages(messages);

                int i = 0,count = 0;
                Message m = snapshot.getValue(Message.class);
                for (Message marr : messages){
                    if(marr.getKey().equals(m.getKey())){
                        count = i;
                    }
                    i++;
                }
                messages.remove(count);
//                messages.add(count,m);
            }

            @Override
            public void onChildMoved(@NonNull DataSnapshot snapshot, @Nullable String previousChildName) {

            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        messages = new ArrayList<>();
    }

    private  void  displayMessages(ArrayList<Message> messages){
        rvMessage.setLayoutManager(new LinearLayoutManager(GroupChatActivity.this));
        adapter = new MessageAdapter(GroupChatActivity.this,messages,reference);
        rvMessage.setAdapter(adapter);

    }
}

File: MyChat/app/src/main/java/com/example/mychat/GroupChatActivity.kt
package com.example.mychat

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.LayoutInflater
import android.widget.Button
import android.widget.EditText
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.google.firebase.storage.FirebaseStorage
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class GroupChatActivity : AppCompatActivity() {
    private lateinit var chatRecyclerView: RecyclerView
    private lateinit var messageBox: EditText
    private lateinit var sendButton: ImageView
    private lateinit var messageAdapter: MessageAdapter
    private lateinit var messageList: ArrayList<Message>
    private lateinit var mDbRef: DatabaseReference
    private lateinit var editDialog: AlertDialog
    private lateinit var groupChatRoomId: String
    private var fileType: String? = null
    private val PICK_IMAGE_VIDEO_REQUEST = 1

    var receiverRoom: String? = null
    var senderRoom: String? = null
    var editMessageInput: EditText? = null
    var groupRoom: String? = null

    @SuppressLint("MissingInflatedId", "WrongViewCast")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_group_chat)
        val name = intent.getStringExtra("name")
        val receiverUid = intent.getStringExtra("uid")
        val senderUid = FirebaseAuth.getInstance().currentUser?.uid
        val sendMediaButton: ImageButton = findViewById(R.id.sendMediaButton)
        sendMediaButton.setOnClickListener {
            openGallery()
        }
        groupChatRoomId = intent.getStringExtra("groupId") ?: ""
        mDbRef = FirebaseDatabase.getInstance().getReference().child("group_chats").child(groupChatRoomId)

        senderRoom = receiverUid + senderUid
        receiverRoom = senderUid + receiverUid
        groupRoom = "messages"



        supportActionBar?.title = name


        chatRecyclerView = findViewById(R.id.chatRecyclerView)
        messageBox = findViewById(R.id.messagebox)
        sendButton = findViewById(R.id.sentButton)

        messageList = ArrayList()
        // Adding delete functionality
        val deleteListener = object : DeleteMessageListener {
            override fun onDeleteMessage(position: Int) {
                deleteMessage(position)
            }
        }
        val editListener = object : EditMessageListener {
            override fun onEditMessage(position: Int) {
                editMessage(position)
            }
        }

        messageAdapter = MessageAdapter(this, messageList, deleteListener, editListener)



        chatRecyclerView.layoutManager = LinearLayoutManager(this)
        chatRecyclerView.adapter = messageAdapter

        //logic for adding data to recycler view
        mDbRef.child(groupRoom!!)
            .addValueEventListener(object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {

                    messageList.clear()

                    for(postSnapshot in snapshot.children){
                        val message = postSnapshot.getValue(Message::class.java)
                        if (message != null) {
                            message.id = postSnapshot.key
                        }
                        messageList.add(message!!)
                    }
                    messageAdapter.notifyDataSetChanged()

                }

                override fun onCancelled(error: DatabaseError) {

                }

            })

        //adding the message to the database
        sendButton.setOnClickListener {

            val message = messageBox.text.toString()
            val dateFormat = SimpleDateFormat("HH:mm:ss", Locale.getDefault())
            val date = Date(System.currentTimeMillis())
            val messageObject = Message(message, senderUid, dateFormat.format(date), "text")

            mDbRef.child(groupRoom!!).push()
                .setValue(messageObject)
            messageBox.setText("")

        }

        // Initialize the edit dialog
        val editView = LayoutInflater.from(this).inflate(R.layout.dialog_edit_message, null)
        editMessageInput = editView.findViewById<EditText>(R.id.editMessageInput)

        editDialog = AlertDialog.Builder(this)
            .setTitle("Edit Message")
            .setView(editView)
            .setPositiveButton("Save") { _, _ ->
                val editedMessage = editMessageInput?.text.toString()
                if (editedMessage.isNotEmpty()) {
                    // Handle saving the edited message to the database
                    saveEditedMessage(selectedMessagePosition, editedMessage)
                }
            }
            .setNegativeButton("Cancel") { dialog, _ ->
                dialog.dismiss()
            }
            .create()
    }

    private fun openGallery() {
        val intent = Intent()
        intent.type = "image/* video/*"
        intent.action = Intent.ACTION_GET_CONTENT
        startActivityForResult(Intent.createChooser(intent, "Select Image or Video"), PICK_IMAGE_VIDEO_REQUEST)
    }
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == PICK_IMAGE_VIDEO_REQUEST && resultCode == RESULT_OK && data != null && data.data != null) {
            val selectedMediaUri: Uri = data.data!!
            val fileType = contentResolver.getType(selectedMediaUri)

            if (fileType?.startsWith("image/") == true || fileType?.startsWith("video/") == true) {
                // Upload the selected image or video to Firebase Storage
                uploadMedia(selectedMediaUri, fileType)
            } else {
                Toast.makeText(this, "Please select a valid image or video", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun uploadMedia(mediaUri: Uri, fileType: String) {
        val storageReference = FirebaseStorage.getInstance().reference
        val mediaRef = storageReference.child("media").child("${System.currentTimeMillis()}.$fileType")
        mediaRef.putFile(mediaUri)
            .addOnSuccessListener { taskSnapshot ->
                // Get the download URL of the uploaded media
                mediaRef.downloadUrl.addOnSuccessListener { uri ->
                    val downloadUrl = uri.toString()
                    // Create a Message object for the uploaded media
                    val dateFormat = SimpleDateFormat("HH:mm:ss", Locale.getDefault())
                    val date = Date(System.currentTimeMillis())
                    val senderUid = FirebaseAuth.getInstance().currentUser?.uid
                    val messageObject = Message(downloadUrl, senderUid, dateFormat.format(date), mediaType = fileType)

                    // Save the message to the database
                    sendMessageToDatabase(messageObject)
                }
            }
            .addOnFailureListener { e ->
                Toast.makeText(this, "Failed to upload media: ${e.message}", Toast.LENGTH_SHORT).show()
            }
    }

    private fun sendMessageToDatabase(message:Message){
        // Save the message to the sender's room
        mDbRef.child(groupRoom!!).push()
            .setValue(message)
            .addOnSuccessListener {

            }
            .addOnFailureListener {
                // Handle failure
            }

    }


    private var selectedMessagePosition: Int = -1
    private fun editMessage(position: Int) {
        selectedMessagePosition = position
        val selectedMessage = messageList[position]
        editDialog.show()
        editMessageInput?.setText(selectedMessage.message)
    }
    private fun saveEditedMessage(position: Int, editedMessage: String) {
        // Handle saving the edited message to the database
        val messageId = messageList[position].id
        if (messageId != null) {
            mDbRef.child(groupRoom!!).child(messageId)
                .child("message").setValue(editedMessage)
                .addOnSuccessListener {
                    // Handle success
                }
                .addOnFailureListener {
                    // Handle failure
                }
        }
    }

    private fun deleteMessage(position: Int) {
        if (position in 0 until messageList.size) {
            val messageId = messageList[position].id // Assuming your Message class has an 'id' property

            // Remove from local list
            messageList.removeAt(position)
            messageAdapter.notifyItemRemoved(position)

            // Remove from the database
            if (messageId != null) {
                mDbRef.child(groupRoom!!).child(messageId)
                    .removeValue().addOnSuccessListener {
                        // Handle success
                    }.addOnFailureListener {
                        // Handle failure
                    }
            }

            if (messageId != null) {
                mDbRef.child(groupRoom!!).child(messageId)
                    .removeValue().addOnSuccessListener {
                        // Handle success
                    }.addOnFailureListener {
                        // Handle failure
                    }
            }
        }
    }

}
